### PR TITLE
Simplify retains grad hook implementation

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92604
* #92734

How the old retains_grad hooks was implemented:
- retains_grad hooks are stored on the autograd_meta, as entries in a vector
- upon registration, a wrapper hook CppFunctionTensorPreHook is created to wrap that vector, and then that wrapper hook is registered to the grad_fn, i.e., by appending it to a vector of retains_grad hooks on the grad_fn 
- upon in-place, for the old grad_fn we set the retains_grad hook to nullptr, so that even though the old grad_fn still references the vector, the vector contains a single nullptr. For the new grad_fn, we create a new wrapper hook around the vector (storing the single retains_grad hook) on autograd_meta.

The new retains_grad hook implementation:
- we store std::function by value, and we store it on the grad_fn rather than the autograd_meta
- a single grad_fn can have multiple outputs, so it can potentially hold multiple retains_grad hooks. We use an unordered_map (previously a vector).
- on in-place we remove the hook from the old grad_fn and put it in the new grad_fn (small implication of this change is that  we we now need to have access to both the old grad_fn and new grad_fn, this isn't a problem)

Other details:
- CppFunctionTensorPreHook took a shared_ptr to vector of std::function. In our new implementation, we add a new wrapper hook CppFunctionSingleTensorPreHook, which takes a single std::function.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
SGTM

File: torch/csrc/autograd/cpp_hook.cpp [hunk 1]
```diff
@@ -48,5 +48,22 @@ variable_list CppFunctionTensorPreHook::operator()(
   return results;
 }
 
+// NOLINTNEXTLINE(modernize-pass-by-value)
+CppFunctionSingleTensorPreHook::CppFunctionSingleTensorPreHook(
+    std::function<at::TensorBase(const at::TensorBase&)> hook,
+    int value_idx)
+    : hook_(hook), value_idx_(value_idx) {}
```
Comment (Skylion007) [inline torch/csrc/autograd/cpp_hook.cpp:55]: @soulitzer Why not just move the hook here? No reason to make an expensive and unnecessary copy?

Comment (soulitzer) [reply torch/csrc/autograd/cpp_hook.cpp:55]: Good catch, I overlooked this one

Comment (Skylion007) [reply torch/csrc/autograd/cpp_hook.cpp:55]: Fixed in my general overhaul of std::move in autograd PR

**Pull request closed without merging**